### PR TITLE
Currency

### LINKED ISSUE #87: [API Request] API to get currency conversion history
Use [Historical currency converter web service](http://currencies.apps.grandtrunk.net/) to get information about currency rates history. Use **List of historical conversion rates** API ([Reference response](http://currencies.apps.grandtrunk.net/getrange/1999-01-01/2000-01-01/eur/usd)).
Currently the API collapses the conversion rates that are same, you need to return a conversion rate for each day.
Do check if from date is before to date. 
Return a 400 response if currency code is incorrect which you can know from API response
### Input Parameters
+ from date (in YYYY-MM-DD format)
+ to date (in YYYY-MM-DD format)
+ from currency code (three digit currency code)
+ to currency code (three digit currency code)
### Output
+ List of currency conversion rates (length of list should be equal to the number of days between to-date and from-date)

### PR DESCRIPTION
**Pull request by darthgera123**
# Description

Returns the api with currency exchange and date 

Fixes #87 (issue)

## Type of change
Just put an x in the [] which are valid.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `flake8`
- [ ] `python manage.py test`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: api/modules/currency/constants.py [hunk 1]
```diff
@@ -3,3 +3,4 @@
 Reference Docs: https://free.currencyconverterapi.com/
 """
 CURRENCY_CONVERTER_API_URL = 'https://free.currencyconverterapi.com/api/v5/convert?q={}'
+CURRENCY_VALUE_DATE_API_URL = 'http://currencies.apps.grandtrunk.net/getrate/{0}/{1}/{2}'
```
Comment (prabhakar267) [inline api/modules/currency/constants.py]: This should be expecting 4 parameters, right?
`start_date`, `end_date`, `source_currency_code`, `target_currency_code`

File: api/urls.py [hunk 2]
```diff
@@ -90,4 +90,5 @@
     # Zomato API
     path('get-all-restaurants/<str:latitude>/<str:longitude>', food_views.get_all_restaurants,
          name="get-all-restaurants"),
+
```
Comment (prabhakar267) [inline api/urls.py]: Add URL

File: api/modules/currency/views.py [hunk 3]
```diff
@@ -33,3 +35,51 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+    :param start date:
+    :param end date:
+    :param source_currency_code:
+    :param target_currency_code:
+    :return 400 response if dates are incorrect
+    :return: 503 if Free Currency Converter api fails
+    :return: 200 successful
+    """
+    currency_list = []
+    y_start = start_date.split('-')[0]
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: Interpreting from variable name `y_start` has YYYY information and in `start = ...` (L57) you are using y_start as `%d`

File: api/modules/currency/views.py [hunk 4]
```diff
@@ -33,3 +35,51 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+    m_start = start_date.split('-')[1]
+    d_start = start_date.split('-')[2]
+    start = datetime.datetime.strptime(y_start+m_start+d_start, '%d%m%Y').date()
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: For 14th July 2018 (`start_date = "14-07-2018"`), you can use:
`datetime.datetime.strptime(start_date, "%d-%m-%Y")`
Not completely sure about the format, but something on these terms, you don't need to split it yourself `strptime` can do it for you
Reference: https://stackoverflow.com/a/466376/5602759

File: api/modules/currency/views.py [hunk 5]
```diff
@@ -33,3 +35,51 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+    y_end = end_date.split('-')[0]
+    m_end = end_date.split('-')[1]
+    d_end = end_date.split('-')[2]
+    end = datetime.datetime.strptime(y_end+m_end+d_end, '%d%m%Y').date()
+    if end < start:
+        error_message = "End Date is before Start Date"
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: `End date should be before start date`

File: api/modules/currency/views.py [hunk 6]
```diff
@@ -33,3 +35,51 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+        if not api_response_start.ok or not api_response_end.ok:
+            error_message = "Incorrect parameters please check dates"
+            return Response(error_message, status=status.HTTP_400_BAD_REQUEST)
+        delta = end - start
+        for day in range(delta.days + 1):
+            actual_date = str(start + timedelta(day))
+            response = requests.get(CURRENCY_VALUE_DATE_API_URL.format(
+                actual_date, source_currency_code, target_currency_code))
+            currency_value = response.json()
+            currency_list.append(CurrencyDate(date=str(actual_date),
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: No need of passing date information
If there are 30 days, a list of 30 should mean currency rates are in order

**Review by prabhakar267 — CHANGES_REQUESTED** [review]
Add a sample request as well to test

File: api/modules/currency/constants.py (hunk 1 above)
Comment (darthgera123) [reply api/modules/currency/constants.py]: Thing is if I use the other URL then it skips through dates which you didn't want. That's why I used a loop

File: api/modules/currency/views.py (hunk 5 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: I'm sorry how should end date be before start date

File: api/modules/currency/views.py (hunk 4 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: thats not working

File: api/modules/currency/views.py (hunk 3 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: done

File: api/urls.py (hunk 2 above)
Comment (darthgera123) [reply api/urls.py]: done

File: api/modules/currency/constants.py (hunk 1 above)
Comment (prabhakar267) [reply api/modules/currency/constants.py]: If the API response is something like this
```
10-07-2018 1.4
12-07-2018 1.5
```
So for 10-07-2018 to 14-07-2018 you can have something like 
```
[
    1.4,
    1.4,
    1.5,
    1.5
    1.5
]
```

Comment (darthgera123) [reply api/modules/currency/constants.py]: But the API request to the server is taking 4 parameters. From the server to the currency API is using 3.

**Review by prabhakar267 — CHANGES_REQUESTED** [review]
`api/get-all-currency-rate/2018-04-20/2018-04-25/INR/USD`
I'm getting a `503` error with message `"tuple index out of range"`

File: api/modules/currency/views.py [hunk 7]
```diff
@@ -33,3 +35,58 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+        error_message = "End Date is before Start Date"
+        return Response(error_message, status=status.HTTP_400_BAD_REQUEST)
+    try:
+        api_response = requests.get(CURRENCY_VALUE_DATE_API_URL.format(
+            start_date, source_currency_code, target_currency_code))
+        if not api_response.ok:
+            error_message = "Incorrect parameters please check dates"
+            return Response(error_message, status=status.HTTP_400_BAD_REQUEST)
+        api_response_content = api_response.content.split('\n')
+        currency_dates = []
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: unnecessary declaration

File: api/modules/currency/views.py [hunk 8]
```diff
@@ -33,3 +35,58 @@ def get_currency_exchange_rate(request, source_currency_code, target_currency_co
...
+                if current_date == next_date:
+                    current_date = next_date
+                    current_value = currency_dates[0][1]
+                    currency_dates = currency_dates[1:]
+
+                if current_date > end_date:
+                    break
+                currency_list.append(CurrencyDate(value=current_value).to_json)
+
+            except Exception as e:
```
Comment (prabhakar267) [inline api/modules/currency/views.py]: Any reason for adding this?

**Review by prabhakar267 — CHANGES_REQUESTED** [review]

File: api/modules/currency/views.py (hunk 7 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: but you only said to check the dates

File: api/modules/currency/views.py (hunk 8 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: reason for adding the class?

File: api/modules/currency/views.py (hunk 7 above)
Comment (prabhakar267) [reply api/modules/currency/views.py]: L66 is not necessary

File: api/modules/currency/views.py (hunk 8 above)
Comment (prabhakar267) [reply api/modules/currency/views.py]: Reason for adding this exception?

File: api/modules/currency/views.py (hunk 7 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: but thats how i define a list right?

File: api/modules/currency/views.py (hunk 8 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: reason being because one gets an index error which is at the end of the list. which can be ignored

Comment (prabhakar267) [reply api/modules/currency/views.py]: Why don't you loop it using a for loop
You anyways know what would be the size of output list
Something like `(end-start).days` can give you number of days

File: api/modules/currency/views.py (hunk 7 above)
Comment (prabhakar267) [reply api/modules/currency/views.py]: You don't need to 
`currency_dates = map(lambda x: x.split(' '), api_response_content)` is doing the magic for you

Comment (darthgera123) [reply api/modules/currency/views.py]: oh sorry yeah i forgot i did that

File: api/modules/currency/views.py (hunk 8 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: okay ill make that change

Comment (darthgera123) [reply api/modules/currency/views.py]: yeah its working without the try except thing

File: api/modules/currency/views.py (hunk 7 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: done

File: api/modules/currency/views.py (hunk 8 above)
Comment (darthgera123) [reply api/modules/currency/views.py]: done

**Review by prabhakar267 — CHANGES_REQUESTED** [review]
`http://localhost:8000/api/get-all-currency-rate/2018-04-20/2018-04-25/INR/USD`
For this request I am getting error

`"tuple index out of range"`

**Comment by darthgera123** [comment]
can you please tell me how do i test as well. Postman cannot do both simultaneously

**Review by prabhakar267 — APPROVED** [review]

**Pull request merged**